File: sdk/go/krasis/collab.go
package krasis

import (
	"encoding/json"
	"fmt"
	"math"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// CollabModule manages WebSocket collaboration sessions.
type CollabModule struct {
	wsBaseURL            string
	token                func() string
	conn                 *websocket.Conn
	noteID               string
	handlers             []CollabHandler
	mu                   sync.Mutex
	reconnectAttempts    int
	maxReconnectAttempts int
	reconnectTimer       *time.Timer
}

// CollabHandler is a callback for collaboration events.
type CollabHandler func(event CollabEvent)

// CollabEvent represents a collaboration event.
type CollabEvent struct {
	Type    string
	Payload map[string]any
	UserID  string
}

// NewCollabModule creates a new collaboration module.
func NewCollabModule(wsBaseURL string, token func() string) *CollabModule {
	return &CollabModule{
		wsBaseURL:            wsBaseURL,
		token:                token,
		maxReconnectAttempts: 5,
	}
}

// Connect establishes a WebSocket connection to a note.
func (c *CollabModule) Connect(noteID string) {
	c.mu.Lock()
	c.noteID = noteID
	c.reconnectAttempts = 0
	c.mu.Unlock()
	c.doConnect()
}

func (c *CollabModule) doConnect() {
	c.mu.Lock()
	if c.conn != nil {
		c.conn.Close()
	}

	url := fmt.Sprintf("%s/ws/collab?note_id=%s&token=%s", c.wsBaseURL, c.noteID, c.token())
	dialer := websocket.Dialer{HandshakeTimeout: 10 * time.Second}
	conn, _, err := dialer.Dial(url, nil)
	if err != nil {
		c.mu.Unlock()
		c.emit(CollabEvent{Type: "error", Payload: map[string]any{"error": err.Error()}})
		c.scheduleReconnect()
		return
	}
	c.conn = conn
	c.mu.Unlock()

	c.emit(CollabEvent{Type: "open"})

	go c.readLoop()
}

func (c *CollabModule) readLoop() {
	for {
		c.mu.Lock()
		conn := c.conn
		c.mu.Unlock()
		if conn == nil {
			return
		}

		_, message, err := conn.ReadMessage()
		if err != nil {
			c.emit(CollabEvent{Type: "close"})
			c.scheduleReconnect()
			return
		}

		var msg struct {
			Type    string         `json:"type"`
			Payload map[string]any `json:"payload"`
			UserID  string         `json:"user_id"`
		}
		if err := json.Unmarshal(message, &msg); err != nil {
			continue
		}

		c.emit(CollabEvent{Type: msg.Type, Payload: msg.Payload, UserID: msg.UserID})
	}
}

// On registers a handler for collaboration events.
func (c *CollabModule) On(handler CollabHandler) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.handlers = append(c.handlers, handler)
}

// SendSync broadcasts a document update.
func (c *CollabModule) SendSync(update string, version int) {
	c.send(map[string]any{
		"type": "sync",
		"payload": map[string]any{
			"update":  update,
			"version": version,
		},
	})
}

// SendAwareness broadcasts awareness state.
func (c *CollabModule) SendAwareness(payload map[string]any) {
	c.send(map[string]any{
		"type":    "awareness",
		"payload": payload,
	})
}

// SendPresenceQuery requests the list of online users.
func (c *CollabModule) SendPresenceQuery() {
	c.send(map[string]any{
		"type":    "awareness_query",
		"payload": map[string]any{},
	})
}

func (c *CollabModule) send(msg map[string]any) {
	c.mu.Lock()
	conn := c.conn
	c.mu.Unlock()
	if conn == nil {
		return
	}
	data, _ := json.Marshal(msg)
	conn.WriteMessage(websocket.TextMessage, data)
}

func (c *CollabModule) emit(event CollabEvent) {
	c.mu.Lock()
	handlers := make([]CollabHandler, len(c.handlers))
	copy(handlers, c.handlers)
	c.mu.Unlock()
	for _, h := range handlers {
		h(event)
	}
}

func (c *CollabModule) scheduleReconnect() {
	c.mu.Lock()
	if c.reconnectAttempts >= c.maxReconnectAttempts {
		c.mu.Unlock()
		return
	}
	c.reconnectAttempts++
	delay := time.Duration(math.Min(float64(1000)*math.Pow(2, float64(c.reconnectAttempts)), 30000)) * time.Millisecond
	c.mu.Unlock()

	c.reconnectTimer = time.AfterFunc(delay, c.doConnect)
}

// Disconnect closes the connection and stops reconnecting.
func (c *CollabModule) Disconnect() {
	if c.reconnectTimer != nil {
		c.reconnectTimer.Stop()
	}
	c.mu.Lock()
	c.reconnectAttempts = c.maxReconnectAttempts
	if c.conn != nil {
		c.conn.Close()
		c.conn = nil
	}
	c.mu.Unlock()
}

File: sdk/python/krasis/collab.py
"""WebSocket collaboration module."""

from __future__ import annotations

import asyncio
import json
import math
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any
from urllib.parse import urlencode

import websockets

from krasis.types import AwarenessPayload, SyncPayload


@dataclass
class CollabEvent:
    type: str = ""
    payload: dict[str, Any] = field(default_factory=dict)
    user_id: str = ""


CollabHandler = Callable[[CollabEvent], None]


class CollabModule:
    """WebSocket collaboration session manager."""

    def __init__(self, ws_base_url: str, token: Callable[[], str]) -> None:
        self._ws_base_url = ws_base_url
        self._token = token
        self._note_id = ""
        self._handlers: list[CollabHandler] = []
        self._ws: websockets.WebSocketClientProtocol | None = None
        self._reconnect_attempts = 0
        self._max_reconnect_attempts = 5
        self._read_task: asyncio.Task[None] | None = None

    def on(self, handler: CollabHandler) -> None:
        """Register a collaboration event handler."""
        self._handlers.append(handler)

    async def connect(self, note_id: str) -> None:
        """Establish a WebSocket connection to a note."""
        self._note_id = note_id
        self._reconnect_attempts = 0
        await self._do_connect()

    async def _do_connect(self) -> None:
        if self._ws:
            await self._ws.close()

        params = urlencode({"note_id": self._note_id, "token": self._token()})
        url = f"{self._ws_base_url}/ws/collab?{params}"

        try:
            self._ws = await websockets.connect(url)
            self._emit(CollabEvent(type="open"))
            self._read_task = asyncio.create_task(self._read_loop())
        except Exception as exc:
            self._emit(CollabEvent(type="error", payload={"error": str(exc)}))
            await self._schedule_reconnect()

    async def _read_loop(self) -> None:
        try:
            assert self._ws is not None
            async for message in self._ws:
                try:
                    msg = json.loads(message)
                    self._emit(CollabEvent(
                        type=msg.get("type", ""),
                        payload=msg.get("payload", {}),
                        user_id=msg.get("user_id", ""),
                    ))
                except json.JSONDecodeError:
                    continue
        except websockets.ConnectionClosed:
            self._emit(CollabEvent(type="close"))
            await self._schedule_reconnect()

    async def send_sync(self, update: str, version: int) -> None:
        """Broadcast a document update."""
        await self._send({
            "type": "sync",
            "payload": {"update": update, "version": version},
        })

    async def send_awareness(self, payload: dict[str, Any]) -> None:
        """Broadcast awareness state."""
        await self._send({"type": "awareness", "payload": payload})

    async def send_presence_query(self) -> None:
        """Request the list of online users."""
        await self._send({"type": "awareness_query", "payload": {}})

    async def _send(self, msg: dict[str, Any]) -> None:
        if self._ws is None:
            return
        await self._ws.send(json.dumps(msg))

    def _emit(self, event: CollabEvent) -> None:
        for handler in self._handlers:
            handler(event)

    async def _schedule_reconnect(self) -> None:
        if self._reconnect_attempts >= self._max_reconnect_attempts:
            return
        self._reconnect_attempts += 1
        delay = min(1.0 * (2 ** self._reconnect_attempts), 30.0)
        await asyncio.sleep(delay)
        await self._do_connect()

    async def disconnect(self) -> None:
        """Close the connection and stop reconnecting."""
        if self._read_task:
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass
        self._reconnect_attempts = self._max_reconnect_attempts
        if self._ws:
            await self._ws.close()
            self._ws = None
